### PR TITLE
SI-6215 Fix compiler crash on private method in value class

### PR DESCRIPTION
**Pull request by adriaanm**
Fixes the problem with private defs in value classes by moving the $extension after the name proper rather than before. The previous scheme did not commute with makeNonPrivate:

I.e. if -ext-> is "generate extension name" and -mnp-> is "make not private" we did get for
method foo in value class Foo:

  foo -ext-> extension$foo -mnp-> Foo$$extension$foo

but

  foo -mnp-> Foo$$foo -ext-> extension$Foo$$foo

With the change both variations give the same name:

  foo -ext-> foo$extension -mnp-> Foo$$foo$extension

but

  foo -mnp-> Foo$$foo -ext-> Foo$$foo$extension

(resubmit of #1427 with fixed commit message)

### CODE REVIEW

**Comment by adriaanm** [comment]
LGTM /cc @gkossakowski

**Comment by gkossakowski** [comment]
No chance for test-case? :-)

**Comment by retronym** [comment]
The [SIP](http://docs.scala-lang.org/sips/pending/value-classes.html) should also be updated to reflect the new expansion scheme.

**Comment by som-snytt** [comment]
I think I see a teeny test case.

I thought this was my chance to say, I wish the issue tracker allowed submitting a pending test case; and then I grabbed the commit; but there was already a test file, strangely enough with the same name as mine...

**Comment by som-snytt** [comment]
Also, thanks to retronym for mentioning at some point that git cherry-pick is a command, not a metaphor.

**Comment by jsuereth** [comment]
LGTM

**Comment by gkossakowski** [comment]
Thanks. Merging.

@retronym: Any chance you could take care of updating SIP?

**Comment by retronym** [comment]
If someone could grant me permissions, I'll edit.

**Comment by gkossakowski** [comment]
@retronym: I'll ping @jsuereth about it.

File: src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala [hunk 1]
```diff
@@ -68,10 +68,10 @@ abstract class ExtensionMethods extends Transform with TypingTransformers {
    */
   def extensionMethod(imeth: Symbol): Symbol = atPhase(currentRun.refchecksPhase) {
     val companionInfo = imeth.owner.companionModule.info
-    val candidates = extensionNames(imeth) map (companionInfo.decl(_))
+    val candidates = extensionNames(imeth) map (companionInfo.decl(_)) filter (_.exists)
     val matching = candidates filter (alt => normalize(alt.tpe, imeth.owner) matches imeth.tpe)
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala:72]: I know these lists are probably small, but in the interest of performance, wouldn't it be better to fuse the two calls to `filter`?

**Pull request merged**